### DISCUSSION TITLE
Questions about SPI

### DISCUSSION THREAD

**Discussion by sheershak-agarwal** [nasa/fprime#1083] [Q&A] [answered]
Hello, 

Situation:
I was working with SPI on F' to connect to my IMU. (I need to run on mode 3 and rx tx bit length is 16, which is quite unusual.)

I have to send one 16 bits command and receive seven 16 bit values. Since RPI doesn't support bit length of 16, so I am sending two eight bits command and receiving fourteen 8 bits values. I understand that there will be 1.5 microsecond delay between the two eight bits and that's what I am trying test with the IMU is good for us.

My problem:
So I am using this code which I took from the RPI example, I can see the clock for the send bytes but not the receive bytes. Please see the diagram attached. So I looked at the SPI driver and in readwrite function https://github.com/nasa/fprime/blob/37b4bfec15837679a138ce87ff82a87ac42a3729/Drv/LinuxSpiDriver/LinuxSpiDriverComponentImpl.cpp#L51, we were just adding length of write buffer.
```
char inBuf[14];
Fw::Buffer in;
in.setData((U8*)inBuf);
in.setSize(sizeof(inBuf));

char buf[2] = {0x3E, 0x00};
Fw::Buffer out;
out.setData((U8*)buf);
out.setSize(sizeof(buf));
this->SpiReadWrite_out(0,out,in);
printf("IMU bytes\n");
for (int i = 0; i < 14; i++) {
  printf("%x ", inBuf[i]);
}
printf("\n");
```
![zoom](https://user-images.githubusercontent.com/80468363/138346363-4b5e4f24-6a53-4fce-8dca-de8f1a258b46.PNG)

So I changed my code to make the write buffer length = write + read length (2+14). and then my data looked a lot better but when I print the bytes on f', they are different from what it is on the image attached.
```
char inBuf[14];
Fw::Buffer in;
in.setData((U8*)inBuf);
in.setSize(sizeof(inBuf));

char buf[16] = {0x00};
buf[0] = 0x3E;
Fw::Buffer out;
out.setData((U8*)buf);
out.setSize(sizeof(buf));
this->SpiReadWrite_out(0,out,in);
printf("IMU bytes\n");
for (int i = 0; i < 16; i++) {
   printf("%x ", inBuf[i]);
}
```
![new 16 = 16](https://user-images.githubusercontent.com/80468363/138349382-2da9303a-1c65-46e0-9694-b8f405b0f6c4.PNG)
The  data  I  am   getting on terminal in this scenario is:
IMU bytes
7f ff 80 8 bf fb bf ea 80 b 80 1 bc 15 bf d0

**Comment by LeStarch** [comment, marked as answer]
My understanding with SPI is that you read and write at the same time.  Thus even if you are only required to write 16 bits, you still need to provide a buffer that is 16bits + 7*16bits long.


What happens if you try this:

```
char inBuf[2 + 14]; // in and out buffers should be the same size
Fw::Buffer in;
in.setData((U8*)inBuf);
in.setSize(sizeof(inBuf));

char buf[2 + 14] = { 0 };
buf[0] = 0x3E;
buf[1] = 0x00;

Fw::Buffer out;
out.setData((U8*)buf);
out.setSize(sizeof(buf));
this->SpiReadWrite_out(0,out,in);
printf("IMU bytes\n");
for (int i = 0; i < 16; i++) {
  printf("%x ", inBuf[i]);  // bytes likely in position 2 - 15, as device needs 16bit command before responding.
}
printf("\n");
```

**Reply by LeStarch** [reply]
This is because sending out bytes provides the clock for reading and writing.

**Reply by sheershak-agarwal** [reply]
input and output buffers should be the same size. That answers my question. Thank you

**Reply by sheershak-agarwal** [reply]
also in the inBuf last 14 bytes will be data received right? what will be first two bytes? same as the write buffer?

**Reply by LeStarch** [reply]
In responding to the update, inBuf should also be 16bytes as 16bytes will be the total received in the transation.  Can you print the output of the print alongside the graph?

**Reply by LeStarch** [reply]
The first two bytes are device dependent.  It will usually say what they should be in the datasheet.  Sometimes the device is fast and can respond with the real data.  Sometimes it will be ack bytes.  Sometimes garbage.

**Reply by sheershak-agarwal** [reply]
yes, i will add that data in a minute. thanks again for the clarification

**Reply by sheershak-agarwal** [reply]
added please check the description

**Reply by LeStarch** [reply]
Can you rule out noise and/or wiring issues between the device and the CPU?  Do you see the data you expect (known bytes/words) on the diagram or in the printed bytes?  Your code seems ok to me...and so I might look for a hardware cause.  Also ensure this:

```
char inBuf[14];
```

Is really this:

```
char inBuf[16];
```

I am assuming that it did not get updated in the description.

**Reply by sheershak-agarwal** [reply]
There is nothing other than IMU connected to my rpi and all the hardware connections are good. I don't know if either of those are the right packets.  I will cross check with other MCU where I know the data is right. Yeah I didn't update the inBuf to 16 in the description but my code has that change.
